### ISSUE TITLE
Equality operator for torch.distribution.*

### ISSUE THREAD

**Issue by Mirko-Nava** [pytorch/pytorch#34949]
## 🚀 Feature
Implement `__eq__()` for torch.distributions.distribution.Distribution and subclasses.

## Motivation

This would avoid situations such as:
```
from torch.distributions import Normal
assert Normal(0, 1) == Normal(0, 1)
AssertionError
```
It should be possible to compare distributions without the need of writing custom code.
The comparison operator could be implemented according to the parameters that define the distribution itself.

As an example, the Normal class could have:
```
def __eq__(self, other):
    return self.loc == other.loc and self.scale == other.scale
```

**Comment by colesbury** [comment]
Thanks for writing up the feature request. I'll let the torch.distributions teams comment.

**Comment by neerajprad** [comment]
Sounds reasonable. One complication that I see is comparing distributions which are differently parameterized. e.g. Bernoulli with probs vs logits. One choice would be to return False if the distribution instances are differently parameterized. We can also convert the parameters to compare, which might be wasteful but more appropriate.

**Comment by fritzo** [comment]
Can you explain more the use case for `==` on distributions? I would't expect e.g. PyTorch modules to be equality comparable. This also seems to bring up tricky issues of e.g. whether two distributions are equal if only `.validate_args` differ. Or what if they differ only in shape, but would be identical if they were broadcast correctly? Do you want equality to behave like tensor equality (which can return a batched result) or boolean equality?

**Comment by alicanb** [comment]
To me, this version makes sense, but i'm not sure if we need it- manually typing this check is more unambiguous and not that cumbersome to type manually 
```python
def __eq__(self, other):
    return kl_divergence(self, other) == 0 
```

**Comment by vishwakftw** [comment]
I was just going to suggest what @alicanb suggested.

**Comment by Mirko-Nava** [comment]
The meaning I give to equality for distributions is related to the underlying formal parameters of the distribution itself (i.e.,  mean and standard deviation for a normal distribution), so my interpretation is boolean valued.
In this sense, adopting the solution suggested by @alicanb seems reasonable (see https://en.wikipedia.org/wiki/Gibbs%27_inequality for details).
The only caveat I could forsee regards differences in shape.
The KL Divergence between a distribution parametrized by real values and one parametrized by tensors should be already working, as broadcasting takes care of it; though the result will be a tensor and one must apply an aggregation function to get a real value.
The problem arises when trying to compute the KL Divergence of distributions parametrized by tensors having different shapes.
That said, I would expect distributions having different shapes to be not equal.
So I guess I would do something like:
```python
def __eq__(self, other):
    if self.batch_shape != other.batch_shape:
        return False
    return kl_divergence(self, other) == 0 
```
Note: wether it should be `batch.shape` or `event_shape` I'm not sure of.

**Comment by fritzo** [comment]
Hi @Mirko-Nava can you please give some example applied use cases?

**Comment by Mirko-Nava** [comment]
I stumbled upon this when writing test cases for a custom layer, for which I tried comparing two distributions to see if they were equal.
Honestly I cannot imagine any other applied use cases other than testing; at least for now.

**Comment by fritzo** [comment]
@Mirko-Nava if your only use case is testing, I would recommend writing a custom test utility to compare distributions. I do this often in test codebases, e.g.
```py
PARAMS = {
    Bernoulli: ("logits",),
    Normal: ("loc", "scale"),
    MultivariateNormal: ("loc", "scale_tril"),
    ...
}
def dist_eq(d1, d2):
    if type(d1) != type(d2):
        return False
    if d1.batch_shape != d2.batch_shape:
        return False
    for name in PARAMS[type(d1)]:
        if not torch.all(getattr(d1, name) == getattr(d2, name)):
            return False
    return True
```

My concern with adding this operator before we have a compelling use case is that there are many design decisions, and we might later find a more compelling use case but realize that we made inconvenient design decisions. So I think it's best to hold off on an expanded interface for now.
